### DISCUSSION TITLE
Getting TestRun in _editor.output

### DISCUSSION THREAD

**Discussion by seemantTUI** [openHPI/codeocean#2677] [Q&A] [answered]
Hello Mr. Serth,

I am working on implementing a feature where users can request feedback from AI. Here’s the use case: after a user clicks the "Score" button, a test is run, and we retrieve the test results. If they wish to receive feedback from ChatGPT, they can click a "Feedback from AI" button. When this button is clicked, the existing test results would be sent to ChatGPT, and the response would appear in a dialog box..

The issue I’m facing is that I am unable to access the testrun or testrun ID in the view where the "Feedback from AI" button is handled. While I could send the results to ChatGPT at the backend when the user clicks "Score," I want to avoid unnecessary API calls. My goal is to only send the test results to ChatGPT when the "Feedback from AI" button is clicked.

Could you guide me on how to pass the testrun or testrun ID to this view, or how I can access it there to send the test results to the ChatGPT API request? Below are screenshots for reference.

Thank You!

<img width="1409" alt="image" src="https://github.com/user-attachments/assets/8b755327-cc83-4892-a7d0-4a1a56088d58">

<img width="1185" alt="image" src="https://github.com/user-attachments/assets/8facda95-fed5-4327-af68-145280407494">

<img width="1454" alt="image" src="https://github.com/user-attachments/assets/2874097e-c84e-4370-876d-1902ea294499">

**Comment by MrSerth** [comment, marked as answer]
Dear @seemantTUI,

many thanks for reaching out again. I fully understand your use case and think it makes sense to support learners with AI-generated feedback.

Regarding the implementation, different approaches can be chosen. Depending on these, the steps to implement the solution might differ. In the current state, the result of the testruns are shared with the frontend (including the actual error messages and the feedback) through the active web socket connection between the browser and the server (`wss://<host>/submissions/<submission ID>/score`). However, this is done before the testruns are persisted (method `SubmissionsController#save_testrun_output`). Hence, when the testruns are sent to the frontend (`client_socket&.send_data(@submission.calculate_score(current_user).to_json)`), we don't have the testrun ID yet. Nevertheless, in the frontend, you have access to the submission ID (which has a 1-to-n relationship to testruns) and the `filename`. The `filename` is included in the JSON response sent through the WebSocket connection and the submission ID is part of response when creating a submission (`createSubmission`, file `submissions.js`; exemplary use in `scoreCode`, file `evaluation.js`) and also used for constructing the required WebSocket URL. With the submission ID and filename at hand, you can fetch the desired testrun in the backend at any time and access all other relevant data. 

If you don't want to mingle with the different ID and filenames, you should also be able to access the latest testrun of a learner for a given exercise. This is easier to implement (and might be sufficient for a spike), but at the cost that potentially a learner using multiple tabs (which is neither recommended nor officially supported anyway) would only get feedback on the current status but not the older version potentially still open.

The view file you referenced is only used as a dummy to populate the scoring and linter cards through JavaScript (function `populateCard` in `editor.js.erb`). Since there is no server-side rendering involved when transforming the Websocket messages into viewable results in the DOM, you cannot simply add the testrun ID to the Slim view. Hence, as mentioned above, you either need to continue passing information within the frontend, or use the other method to access the latest testrun I outlined above. 

Either way, you would achieve the desired result and only perform an API call to ChatGPT if initiated by the user.

Good luck with the implementation; looking forward to see your results!

Best,
Sebastian
